realm/realm-library/src/androidTest/java/io/realm/RealmListTests.java

```diff
@@ -1019,4 +1022,213 @@ public void testRemoveAllFromRealmInvalidListShouldThrow() {
             assertEquals("This Realm instance has already been closed, making it unusable.", e.getMessage());
         }
     }
+
+    private interface ListModifier <T extends RealmObject> {
+        void modify(RealmList<T> list, T element);
+    }
+
+    private void doTestModifyListFromOtherThread(final ListModifier<Dog> modifier) {
+        final CountDownLatch finishedLatch = new CountDownLatch(1);
+        final Dog dog = testRealm.where(Dog.class).findFirst();
+
+        new Thread(new Runnable() {
+            @Override
+            public void run() {
+                Realm realm = Realm.getInstance(testRealm.getConfiguration());
+                realm.beginTransaction();
+                RealmList<Dog> list = realm.createObject(Owner.class).getDogs();
+                list.add(realm.createObject(Dog.class));
+                try {
+                    modifier.modify(list, dog);
+                    fail();
+                } catch (IllegalArgumentException expected) {
+                    assertEquals("Cannot copy an object from another Realm instance.", expected.getMessage());
+                }
+                realm.cancelTransaction();
+                realm.close();
+                finishedLatch.countDown();
+            }
+        }).start();
+        TestHelper.awaitOrFail(finishedLatch);
+    }
+
+    @Test
+    public void add_objectFromOtherThread() {
+        doTestModifyListFromOtherThread(new ListModifier<Dog>() {
+            @Override
+            public void modify(RealmList<Dog> list, Dog element) {
+                list.add(element);
+            }
+        });
+    }
+
+    @Test
+    public void add_withLocationObjectFromOtherThread() {
+        doTestModifyListFromOtherThread(new ListModifier<Dog>() {
```

Comment:
In e.g. RealmResults all these tests are combined in one method: `runMethodOnWrongThread`. It takes up a lot less space than this? Not sure what is better, but I probably lean slightly towards what RealmResultsTests are doing?